### PR TITLE
IAST destructor intrusive list

### PR DESCRIPTION
**Pull request by kitaisreal**
### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

### CODE REVIEW

**Review by alexey-milovidov — APPROVED** [review]

**Comment by alexey-milovidov** [comment]
What was the purpose of the original code?
It looks like some madness that should be removed.

**Comment by KochetovNicolai** [comment]
@alexey-milovidov you've seen the original pr #39956. If you think I was mad you should have told this earlier.

I thought it was clear enough. We avoid recursive destructor call which can lead to stack overflow.

**Comment by kitaisreal** [comment]
I think original pull request make sense https://github.com/ClickHouse/ClickHouse/pull/39956, because it could be potentially easy to trigger such scenario (same should be done in Analyzer for query tree node https://github.com/ClickHouse/ClickHouse/pull/31796/files#diff-d2848e68e5570fccca705d81aa491f32e494a769de9129062adce497bc952fc5R54).
It probably make sense additionally to merge this pull request to avoid allocation, althought code becomes even more complicated then before.

File: src/Parsers/IAST.cpp [hunk 1]
```diff
@@ -28,34 +28,86 @@ const char * IAST::hilite_none         = "\033[0m";
...
 IAST::~IAST()
 {
-    /// If deleter was set, move our children to it.
-    /// Will avoid recursive destruction.
-    if (deleter)
+    /** Create intrusive linked list of children to delete.
+      * Each ASTPtr child contains pointer to next child to delete.
+      */
+    ASTPtr delete_list_head_holder = nullptr;
+    bool delete_directly = next_to_delete_list_head == nullptr;
```
Comment (davenger) [inline src/Parsers/IAST.cpp:35]: Looks like this can be `const bool` just for safety

File: src/Parsers/IAST.cpp [hunk 2]
```diff
@@ -28,34 +28,86 @@ const char * IAST::hilite_none         = "\033[0m";
...
+          */
+        if (delete_list_head_reference == nullptr)
+        {
+            delete_list_head_reference = std::move(next_node_to_delete);
+        }
+        else
+        {
+            ASTPtr delete_list_tail = delete_list_head_reference;
+
+            while (delete_list_tail)
```
Comment (davenger) [inline src/Parsers/IAST.cpp:100]: A small suggestion how to make this loop a bit more concise and with only one check

```
/// Find the end of the list
while (delete_list_tail->next_to_delete != nullptr)
    delete_list_tail = delete_list_tail->next_to_delete;
/// Append next_node_to_delete to the end
delete_list_tail->next_to_delete = std::move(next_node_to_delete);
```

**Review by davenger — APPROVED** [review]
LGTM
Just a couple of small suggestions, no need to address them in this PR.

File: src/Parsers/IAST.cpp [hunk 3]
```diff
@@ -28,34 +28,86 @@ const char * IAST::hilite_none         = "\033[0m";
...
+          * If we try to move child to delete children into list before subclasses desruction,
+          * first child use count will be 2.
+          */
         {
-            /// If two threads remove ASTPtr concurrently,
-            /// it is possible that neither thead will see use_count == 1.
-            /// It is ok. Will need one more extra stack frame in this case.
-            if (node.use_count() == 1)
+            ASTPtr child_to_delete;
+            child_to_delete.swap(delete_list_head_reference);
```
Comment (KochetovNicolai) [inline src/Parsers/IAST.cpp:84]: I did not understand why we have to swap here (and then traverse a list again).
Can't we just set `child_to_delete->next_to_delete_list_head` to `&next_node_to_delete`? And use itself only if next_node_to_delete is empty.

Comment (kitaisreal) [reply src/Parsers/IAST.cpp]: Good catch. I did not change this place after I changed list append strategy, before I passed tail and added to tail, now as we replace list head with new list head and point it to previous head, we can safely do this.
We can write it like this:
```
ASTPtr child_to_delete;
child_to_delete.swap(delete_list_head_reference);

delete_list_head_reference = std::move(child_to_delete->next_to_delete);
child_to_delete->next_to_delete_list_head = &delete_list_head_reference;
```
Related to swap, I actually want to be more explicit about what is going on here. Updated comments.

File: src/Parsers/IAST.cpp (hunk 1 above)
Comment (kitaisreal) [reply src/Parsers/IAST.cpp]: Agreed. Updated.

**Comment by kitaisreal** [comment]
Tests failures look unrelated.

**Pull request merged**